### DISCUSSION TITLE
Using HIP-backend on Windows

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1240] [Q&A]
Now that HIP is officially supported on Windows, can AdaptiveCpp be built to use it as a backend?

**Comment by blinkfrog** [comment]
I've just tried to built AdaptiveCpp with HIP support. First, I rebuilt LLVM with HIP support by adding `AMDGPU` to `LLVM_TARGETS_TO_BUILD`. AdaptiveCpp, still, refused to build until I added also `-DLLVM_ENABLE_RUNTIMES="compiler-rt"`` to compiler line (CUDA didn't require this). After this, and after adding `C:\Program Files\AMD\ROCm\5.5\bin` to `PATH` environment variable, AdaptiveCpp found HIP and was successfully built.

However, when trying to compile any SYCL-program, I get this error:
`acpp fatal error: Required command line argument --acpp-rocm-path or environment variable ACPP_ROCM_PATH not specified`.

If I add `--acpp-rocm-path="C:/Program Files/AMD/ROCm/5.5"`, it tries to compile, but I get these errors during the linking stage:
```
LINK : warning LNK4044: unrecognized option '/rpath=C:/Program Files/AMD/ROCm/5.5/lib'; ignored
LINK : warning LNK4044: unrecognized option '/rpath=C:/Program Files/AMD/ROCm/5.5/hip/lib'; ignored
LINK : fatal error LNK1181: cannot open input file 'amdhip64.lib'
```
`C:/Program Files/AMD/ROCm/5.5/lib` is correct path (and `C:/Program Files/AMD/ROCm/5.5/hip/lib` is incorrect), but these options are ignored anyway.

So, it can be some error (if needed, I can open an issue).

I workarounded it by copying this lib to current folder (alternatively it is possible to set it in `target_link_directories` when using cmake), and it compiled, but I currently can't test if it works as I don't have AMD GPU (will try to test on another computer).

Also, a bit strange, is that when I tried to compile this as cmake project, `ACPP_ROCM_PATH` variable was ignored, and only `--acpp-rocm-path` in `target_compile_options` worked.

**Reply by fodinabor** [reply]
Cool, thanks for trying :)
I don't think I'll be able to test this, either, as my AMD GPU workstation is currently hogged...

Regarding CMake ignoring the options: I think we just have a few `if(not WIN32)` statements inside CMake since HIP never was supported on Windows (and ROCm as a runtime itself still isn't, no?). Might be worth trying to remove those... and also specialize the command lines - e.g. Windows doesn't know about rpath, so one would just not add that option to the link flags..

**Reply by blinkfrog** [reply]
I think ROCm runtime is supported now on Windows. According to this page: ~https://rocm.docs.amd.com/en/latest/release/windows_support.html~
Edit: now it is this page: https://rocm.docs.amd.com/projects/install-on-windows/en/latest/reference/system-requirements.html

>Support on Windows is provided with two levels on enablement.
>Runtime: Runtime enables the use of the HIP/OpenCL runtimes only.
>HIP SDK: Runtime plus additional components refer to libraries found under [Math Libraries] ...

I'll try to find computers with AMD GPUs to test the executable I compiled.

I also want to ask if it is possible to compile universal binary that would support CUDA and HIP backends and work with GPU which is available. Currently, exe with HIP-backend which I compiled is crashed with System Error saying that amdhip64.dll was not found (it is shipped with AMD drivers), and this error occurs before I can catch it with try/catch. So, an universal binary with CUDA and HIP support probably won't work on a system which has nVidia card but hasn't AMD card.

**Reply by fodinabor** [reply]
Cool, thanks for the link :)

hmm... in theory, the executable should not link directly to the `amdhip64.lib` nor the `amdhip64.dll`, if I'm not mistaken.. that's the whole reason why we have a plugin architecture with different backend DLLs / `.so`s -> trying to load the library should fail gracefully if a required library isn't found... so 2 things:
- @illuhad I am confused, why are we linking the user application with `cudart` and `amdhip64` (in (explicit) multipass)?
- I'm not sure the APIs that we use on Windows for the dynamic loading are indeed fault-tolerant ... can you tell (e.g. by setting the env variable `HIPSYCL_DEBUG_LEVEL=3`) if it's immediately at exe start or if it prints some diagnostics of trying to load backend plugins first?

**Reply by illuhad** [reply]
>     @illuhad I am confused, why are we linking the user application with cudart and amdhip64 (in (explicit) multipass)?

Probably mostly due to historic reasons when we didn't have the plugin architecture yet. It might be necessary for some interop stuff, so perhaps removing them changes interop guarantees. I'm also not sure what would happen to our CUDA/HIP kernel invocations (which ultimately rely on `hip/cudaLaunchKernel`) inside the kernel launchers, which get compiled as part of application code. So, it might be worth investigating if they can be removed, but I'm not sure whether this is indeed the case.

**Reply by blinkfrog** [reply]
@fodinabor, unfortunately, compiled app is crashed with that System error ("The code execution cannot proceed because `amdhip64.dll` was not found. Reinstalling the program may fix this problem") before any debug message at any debug level is printed.

I think, this problem could be managed if `LoadLibrary` and `GetProcAddress` would be used to load `amdhip64.dll` at runtime instead of linking against `amdhip64.lib` at build time, but I don't know if AdaptiveCpp architecture allows this.

**Reply by blinkfrog** [reply]
Also, this is interesting, now that I have rt-backend-hip.dll plugin, and launch a binary which was compiled without any hip targets (for example, it targets cuda and/or omp), I get this warning:
`[AdaptiveCpp Warning] backend_loader: Could not load backend plugin: C:/AdaptiveCpp\bin\hipSYCL\rt-backend-hip.dll with: 126`
Probably, this is because of the same problem: this plugin requires `amdhip64.dll`. Still, this isn't clear why AdaptiveCpp can't just output the same warning instead of crashing when `hip` is one of targets.

**Comment by blinkfrog** [comment]
So, now I've had an opportunity to test my binary on a computer with AMD videocard. And it works! So, we can now say that AdaptiveCpp supports HIP-backends on Windows. Great!

There are some caveats though:

1. That uncatchable and not-processible `The code execution cannot proceed because amdhip64.dll was not found. Reinstalling the program may fix this problem` error still remains on a computers without AMD-card (and drivers) installed, meaning that this isn't possible currently to compile a single AdaptiveCpp binary with support of `cuda` and `hip`-backends - current workaround I use is to compile them into a separate dlls and load required dll dynamically. The opposite thing (trying to run this binary on a system with AMD-card but without nVidia-card installed) works, although still requires the presence of `cudart64_12.dll` file, causing uncatchable error otherwise.
Edit: probably, if I would put `amdhip64.dll` into the directory with acpp binary, this would solve this problem (same way as having `cudart64_12.dll` in the directory of acpp binary solves the opposite problem), but I can't test this as I have no `amdhip64.dll` on my computer.

3. `--acpp-rocm-path` parameter, as well as `ACPP_ROCM_PATH` environment variable don't work, causing `Warning		ignoring unknown argument '-rpath=<HIP_PATH>` warning and following linking error. This currently can be solved by copying `amdhip64.lib` into build directory.

**Reply by illuhad** [reply]
Very cool, thanks for the feedback! We should perhaps extend the wiki page on using AdaptiveCpp on Windows: https://github.com/AdaptiveCpp/AdaptiveCpp/wiki/Using-AdaptiveCpp-on-Windows

For 1: Our runtime library already has a modular plugin system for backends, and it should simply skip a backend if it cannot be loaded. I believe this should also work on Windows (@fodinabor, is there a known limitation on Windows?).
I suspect the issue you are seeing is because for `--acpp-targets=hip` or `--acpp-targets=cuda`, `acpp` uses additional linker flags which causes the application to be explicitly linked against ROCm/CUDA:
https://github.com/AdaptiveCpp/AdaptiveCpp/blob/57b7202983748117b5a3c52437a5fece9cece001/CMakeLists.txt#L389

I'm not sure whether we need this behavior still with recent AdaptiveCpp. I believe we only do this for historic reasons since early versions of AdaptiveCpp had a different runtime library architecture (or rather, lack thereof ;) ). But this is no longer the case. As an additional data point, the generic SSCP compiler also does not directly link the user application against backend runtime libraries and there it works fine.

I would suggest to try to adapt the cmake variables `CUDA_LINK_LINE` and `ROCM_LINK_LINE` when building AdaptiveCpp and remove the explicit linking to backend runtime libraries there. This will likely solve your problem.

We should consider whether we want to also make this change upstream, but it could be a breaking change for anybody who does CUDA/HIP interop since these libraries would then need to be linked by users.

It would be unfortunate if users had to recreate a multi-backend shared library system which our runtime in principle already provides :(

For 2: Windows linker does not seem to support `-rpath` flag; The solution is the same - edit `ROCM_LINK_LINE` and remove the `-Wl,-rpath=...` part :)

**Reply by fodinabor** [reply]
Jup, would be my intuition as well that the issue is that the client application is also linked against the `amdhip.dll`..
The `dlopen` equivalent should also just return an error upon not finding a lib and not open a dialogue.

**Reply by blinkfrog** [reply]
Thank you @illuhad, I've tried what you suggested, and it partly helped.

But, first, I want to note that I forgot to write about additional change that was required to make to make any binary compilable with HIP-target. Some time ago, I [made a PR](https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1264), which solved compilation issue with CUDA-backend by adding `cmath` file, where `double`-overload of `fma` was declared. While this worked for CUDA, it still didn't work for HIP: additionally to `device` attribute, `host` attribute should be added to this declaration. After adding `host` attribute, app started compiling with HIP-backend. If this would be useful, I can create corresponding PR.

Now, to the changes I made. I replaced that ROCM_LINK_LINE setting by the next lines:
```
set(DEFAULT_WIN32_HIP_LINK_LINE "-L$HIPSYCL_ROCM_PATH/lib -lamdhip64")
if(WIN32)
  if(NOT ROCM_LINK_LINE)
    set(ROCM_LINK_LINE ${DEFAULT_WIN32_HIP_LINK_LINE} CACHE STRING "Arguments passed to compiler to link ROCm libraries to SYCL applications")
  endif()
else()
  if(NOT ROCM_LINK_LINE)
    set(ROCM_LINK_LINE "-Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib ${ROCM_LIBS}" CACHE STRING "Arguments passed to compiler to link ROCm libraries to SYCL applications")
  endif()
endif()
```

And this helped to compile programs which target HIP-backend: it isn't needed to place `amdhip64.lib` into the build directory anymore (If this helps, I can create corresponding PR).

However, this still haven't solved these uncatchable errors, occurring when `cudart64_12.dll` is missing on the computer when binary was compiled for CUDA-backend, or  `amdhip64.dll` is missing when HIP was targeted.

Usually, such errors occur on Windows when an app uses implicit dynamic linking against a lib which has corresponding dll pair. In this case, Windows manages the process of loading of this dll, and, in this case, it is almost impossible to handle cases when this dll cannot be loaded because it, or its dependencies are missing (the only ways I am aware of are to use deferred loading which is MSVC-only thing, or SEH).

When using explicit dynamic linking (using `LoadLibrary` and `GetProcAddress` functions), dll loading is fully controllable and allows to handle situations when dll can't be loaded.

So, it looks like AdaptiveCpp still links user applications against CUDA/HIP dlls using implicit dynamic linking at compile-time, as @fodinabor suggests.

**Reply by blinkfrog** [reply]
Just to test, I ran a dryrun, analyzed commands I got and removed `-lamdhip64` flag from the linking command, to prevent app from having a hard dependency on `amdhip64.dll`, which loading is managed by Windows and causes that uncatchable error. This didn't work, I got a lot of missing symbols errors,  indicating that these symbols were expected to be resolved through the HIP runtime library at compile-time. It looks like that AdaptiveCpp still relies on compile-time linking with the HIP (or CUDA) libraries. This contradicts the behavior @illuhad and @fodinabor mentioned, where backend libraries like `amdhip64.dll` would be loaded dynamically at runtime, similar to how dlopen works on Linux (or `LoadLibrary` on Windows).

Still, may be I just build AdaptiveCpp incorrectly? I use the next command to configure it:

`cmake .. -G Ninja -DCMAKE_C_COMPILER=c:/llvm/bin/clang.exe -DCMAKE_CXX_COMPILER=c:/llvm/bin/clang++.exe -DBOOST_ROOT=c:/libraries/boost_1_83_clangw17 -DLLVM_DIR=c:/llvm/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=c:/AdaptiveCpp/ -DWITH_SSCP_COMPILER=OFF -DWITH_OPENCL_BACKEND=OFF -DWITH_LEVEL_ZERO_BACKEND=OFF -DWITH_ACCELERATED_CPU=ON`

**Reply by fodinabor** [reply]
Mhh.. Can you run `ninja -v`, take the linking command line that is output and run that manually with an added `-v` to get the full command line that is passed to the linker?

Also, you may check the DLL dependencies of the executable using https://github.com/lucasg/Dependencies for example, to see if it's possibly an implicit dependency of another lib?

**Reply by blinkfrog** [reply]
Sure!

Here is the linking command line:

```
[2/2] cmd.exe /C "cd . &&  python C:/AdaptiveCpp/lib/cmake/AdaptiveCpp/syclcc-launcher --launcher-cxx-compiler=C:/llvm/bin/clang++.exe --launcher-syclcc="python*C:/AdaptiveCpp/bin/acpp"  --acpp-targets="omp;hip:gfx1030" C:\llvm\bin\clang++.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/Blur.dir/blur.cpp.obj -o Blur.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:Blur.lib -Xlinker /pdb:Blur.pdb -Xlinker /version:0.0   C:/AdaptiveCpp/lib/acpp-rt.lib  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
```

And here is its output when I run it with `-v` flag added to `clang++` invocation for linking:

```
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\llvm\bin
Found HIP installation: C:\Program Files\AMD\ROCm\5.5\, version 5.5.0
 "C:\\llvm\\bin\\lld-link" -out:Blur.exe "-libpath:C:\\llvm\\lib\\clang\\17\\lib\\windows" "-libpath:C:\\AdaptiveCpp\\bin\\..\\lib/" -libpath:C:/Libraries/boost_1_83_clangw17/lib "-libpath:C:\\Program Files\\AMD\\ROCm\\5.5\\/lib" -nologo -nodefaultlib:vcomp.lib -nodefaultlib:vcompd.lib "-libpath:C:\\llvm\\bin/../lib" -defaultlib:libomp.lib /subsystem:console CMakeFiles/Blur.dir/blur.cpp.obj /MANIFEST:EMBED /implib:Blur.lib /pdb:Blur.pdb /version:0.0 C:/AdaptiveCpp/lib/acpp-rt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib oldnames.lib acpp-rt.lib libboost_context-clangw17-mt-x64-1_83.lib libboost_fiber-clangw17-mt-x64-1_83.lib amdhip64.lib
 ```

**Reply by illuhad** [reply]
> This didn't work, I got a lot of missing symbols errors, indicating that these symbols were expected to be resolved through the HIP runtime library at compile-time.

Which symbols? The only thing I can imagine is `hipLaunchKernel` and `__hipPushCallConfiguration`, which currently might be emitted by clang as part of the kernel invocation.

**Reply by blinkfrog** [reply]
> Which symbols? The only thing I can imagine is `hipLaunchKernel` and `__hipPushCallConfiguration`, which currently might be emitted by clang as part of the kernel invocation.

`hipLaunchKernel`, `__hipPushCallConfiguration`, `__hipPopCallConfiguration`, `__hipRegisterFatBinary`, `__hipRegisterFunction`, `__hipUnregisterFatBinary`.

**Reply by illuhad** [reply]
I think the `__hip(Un)Register*` functions you can perhaps resolve by putting both backends into explicit multipass mode: `cuda.explicit-multipass:...;hip.explicit-multipass:...`. The others are related to kernel launch as I said.

**Reply by blinkfrog** [reply]
Wow, this helped! Thank you very much! I now can compile a single binary with omp, cuda and amd backends on Windows, and there is no runtime errors! Amazing!

Could you make changes to AdaptiveCpp's `cmakelists.txt`, similar to I made:
```
set(DEFAULT_WIN32_HIP_LINK_LINE "-L$HIPSYCL_ROCM_PATH/lib -lamdhip64")
if(WIN32)
  if(NOT ROCM_LINK_LINE)
    set(ROCM_LINK_LINE ${DEFAULT_WIN32_HIP_LINK_LINE} CACHE STRING "Arguments passed to compiler to link ROCm libraries to SYCL applications")
  endif()
else()
  if(NOT ROCM_LINK_LINE)
    set(ROCM_LINK_LINE "-Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib ${ROCM_LIBS}" CACHE STRING "Arguments passed to compiler to link ROCm libraries to SYCL applications")
  endif()
endif()
```
or, alternatively, would you consider my PR with these changes if I make it? The same question is regarding adding `host` attribute to `fma` declaration in [cmath](https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/include/hipSYCL/std/hiplike/cmath) file?

**Reply by illuhad** [reply]
Very cool! I totally forgot that the multipass modes also have different kernel invocation logic, so the dependencies on `hipLaunchKernel` are gone too :) (Seems like we were clever when designing this mode ;) )

I'm not sure we can globally enable the change to CMakeLists, as the default integrated multipass modes will still need the additional linking steps. We would have to distinguish between an integrated and an explicit multipass link line. Users today already can enforce the behavior that you want by building AdaptiveCpp using `cmake -DROCM_LINK_LINE=" ... "`.

> The same question is regarding adding host attribute to fma declaration in [cmath](https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/include/hipSYCL/std/hiplike/cmath) file?

This change seems more uncritical. Submit a PR and if everything in CI is fine, we can merge it :)

**Reply by blinkfrog** [reply]
>I'm not sure we can globally enable the change to CMakeLists, as the default integrated multipass modes will still need the additional linking steps.

But that proposed change is conditional, on Linux and Mac all remains the same, it's just `-rpath` removed on Windows, where it doesn't work anyway, causing link-time errors which can be avoided only by placing `amdhip64.lib` into to build directory.

Well, with using `explicit-multipass`, it probably will compile without this change, not sure. But this change also make possible to compile binaries with integrated multipass, let they are hard-dependent on `amdhip64.dll`.

**Reply by illuhad** [reply]
Sorry, I missed that you are just removing rpath on Windows. In that case, it might be fine. Can you open a PR so that we can have CI check it?

**Reply by blinkfrog** [reply]
Thanks, I've opened two PRs. :)

**Comment by blinkfrog** [comment]
Closing the discussion, as all problems were solved. HIP-backend is proved to work on Windows, and, when using `explicit-multipass`, all issues with dependencies are resolved: it is now possible to compile a single binary which would target OMP, HIP and CUDA and could work on a computer which has no GPUs, supporting CUDA or/and HIP without any runtime errors.

**Comment by blinkfrog** [comment]
I am reopening the discussion.

HIP backend (when using HIP SDK 5.5) works on Windows when using integrated multipass compile flow.
However, when I compile binaries using explicit multipass, it doesn't. It correctly finds the device, creates queue, but can't perform kernels.

This is the output I get:

```
[AdaptiveCpp Warning] backend_loader: Could not load library: g:\Temp\amdtest\hipSYCL\rt-backend-cuda.dll with: 126
Running on device: AMD Radeon RX 7600
[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/hip/hip_code_object.cpp:67 @ build_hip_module(): hip_executable_object: could not create module (error code = HIP:218)
[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/hip/hip_queue.cpp:574 @ submit_multipass_kernel_from_code_object(): hip_queue: Code object construction failed
[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/hip/hip_code_object.cpp:67 @ build_hip_module(): hip_executable_object: could not create module (error code = HIP:218)
[AdaptiveCpp Error] from D:/source/AdaptiveCpp/src/runtime/hip/hip_queue.cpp:574 @ submit_multipass_kernel_from_code_object(): hip_queue: Code object construction failed
============== hipSYCL error report ==============
hipSYCL has caught the following undhandled asynchronous errors:
   0. from D:/source/AdaptiveCpp/src/runtime/hip/hip_code_object.cpp:67 @ build_hip_module(): hip_executable_object: could not create module (error code = HIP:218)
   1. from D:/source/AdaptiveCpp/src/runtime/hip/hip_queue.cpp:574 @ submit_multipass_kernel_from_code_object(): hip_queue: Code object construction failed
   2. from D:/source/AdaptiveCpp/src/runtime/hip/hip_code_object.cpp:67 @ build_hip_module(): hip_executable_object: could not create module (error code = HIP:218)
   3. from D:/source/AdaptiveCpp/src/runtime/hip/hip_queue.cpp:574 @ submit_multipass_kernel_from_code_object(): hip_queue: Code object construction failed
The application will now be terminated.
```

The problem is only with `hip.explicit-multipass`; `cuda.explicit-multipass` works well, and so is `hip.integrated-multipass`.

**Comment by blinkfrog** [comment]
I am closing this discussion as I decided to move the last mentioned problem into a separate discussion, so it would be easier to find it using search: https://github.com/AdaptiveCpp/AdaptiveCpp/issues/1447
